### PR TITLE
Add mempool filter

### PR DESCRIPTION
**Pull request by sdschmidt**
## Pull request checklist

- [x] All tests pass. Demo project builds and runs.
- [x] I have resolved any merge conflicts.

### CODE REVIEW

File: collections/src/limit_hash_set.rs [hunk 1]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (paberr) [inline collections/src/limit_hash_set.rs]: Remove unnecessary empty line at the beginning

File: mempool/Cargo.toml [hunk 2]
```diff
@@ -13,6 +13,8 @@ travis-ci = { repository = "https://github.com/nimiq/core-rs", branch = "master"
 maintenance = { status = "experimental" }
 
 [dependencies]
+serde = "1.0.85"
+serde_derive = "1.0.85"
```
Comment (paberr) [inline mempool/Cargo.toml]: Remove `serde` dependencies.
```suggestion
```

File: mempool/Cargo.toml [hunk 3]
```diff
@@ -13,6 +13,8 @@ travis-ci = { repository = "https://github.com/nimiq/core-rs", branch = "master"
 maintenance = { status = "experimental" }
 
 [dependencies]
+serde = "1.0.85"
```
Comment (paberr) [inline mempool/Cargo.toml]: Remove `serde` dependencies.
```suggestion
```

File: mempool/src/filter.rs [hunk 4]
```diff
@@ -0,0 +1,155 @@
+use account::Account;
+use collections::LimitHashSet;
+use keys::Address;
+use nimiq_hash::{Blake2bHash, Hash};
+use primitives::coin::Coin;
+use primitives::networks::NetworkId;
+use transaction::{Transaction, TransactionFlags};
+
+
```
Comment (paberr) [inline mempool/src/filter.rs]: 1-2 newlines suffice.

File: mempool/src/filter.rs [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+        creation_value: Coin::ZERO,
+        sender_balance: Coin::ZERO,
+        recipient_balance: Coin::ZERO,
+    };
+
+    pub const DEFAULT_BLACKLIST_SIZE: usize = 25000;
+
+    pub fn new(rules: Rules, blacklist_limit: usize) -> Self {
+        let b: LimitHashSet<Blake2bHash> = LimitHashSet::new(blacklist_limit);
+        MempoolFilter {blacklist: b, rules}
```
Comment (paberr) [inline mempool/src/filter.rs]: Formatting
```suggestion
        MempoolFilter {
            blacklist: b,
            rules,
        }
```

File: mempool/src/filter.rs [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+
+    pub fn accepts_transaction(& self, tx: &Transaction) -> bool {
+         if tx.fee < self.rules.tx_fee ||
+             tx.value < self.rules.tx_value ||
+             tx.value + tx.fee < self.rules.tx_value_total ||
+             tx.fee_per_byte() < self.rules.tx_fee_per_byte {
+             return false;
+         } else {
+             match tx.flags {
+                 TransactionFlags::CONTRACT_CREATION => {
```
Comment (paberr) [inline mempool/src/filter.rs]: `TransactionFlags` is a bitflag. This match will only evaluate to true if `CONTRACT_CREATION` is the *only* flag set.
If you want to check whether `CONTRACT_CREATION` is one of the flags, check:
`tx.flags.contains(TransactionFlags::CONTRACT_CREATION)`

File: mempool/src/lib.rs [hunk 7]
```diff
@@ -36,6 +40,7 @@ struct MempoolState {
     transactions_by_sender: HashMap<Address, BTreeSet<Arc<Transaction>>>,
     transactions_by_recipient: HashMap<Address, BTreeSet<Arc<Transaction>>>,
     transactions_sorted_fee: BTreeSet<Arc<Transaction>>, // sorted by fee, ascending
+    filter: filter::MempoolFilter,
```
Comment (paberr) [inline mempool/src/lib.rs]: Use `use crate::filter::MempoolFilter;` to improve readability.
That should let you remove all the `filter::` prefixes in this file.

File: mempool/src/lib.rs [hunk 8]
```diff
@@ -56,6 +61,8 @@ impl<'env> Mempool<'env> {
                 transactions_by_sender: HashMap::new(),
                 transactions_by_recipient: HashMap::new(),
                 transactions_sorted_fee: BTreeSet::new(),
+                filter: filter::MempoolFilter::new(filter::MempoolFilter::DEFAULT_RULES,
```
Comment (paberr) [inline mempool/src/lib.rs]: Implement `Default` for `MempoolFilter` and change to:
```suggestion
                filter: MempoolFilter::default(),
```

File: mempool/src/lib.rs [hunk 9]
```diff
@@ -77,6 +84,12 @@ impl<'env> Mempool<'env> {
         {
             let state = self.state.read();
 
+            // Check transaction against rules and blacklist
+            if !state.filter.accepts_transaction(&transaction) || state.filter.blacklisted(&transaction){
```
Comment (paberr) [inline mempool/src/lib.rs]: Missing space before `{`.
```suggestion
            if !state.filter.accepts_transaction(&transaction) || state.filter.blacklisted(&transaction) {
```

File: mempool/src/filter.rs [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+        self.blacklist.remove(&hash);
+        self
+    }
+
+    pub fn blacklisted(&self, tx: &Transaction) -> bool {
+        let hash: Blake2bHash = tx.hash();
+        self.blacklist.contains(&hash)
+    }
+
+    pub fn accepts_transaction(& self, tx: &Transaction) -> bool {
```
Comment (paberr) [inline mempool/src/filter.rs]: Unnecessary space between `&` and `self`.
```suggestion
    pub fn accepts_transaction(&self, tx: &Transaction) -> bool {
```

File: mempool/src/lib.rs [hunk 11]
```diff
@@ -77,6 +84,12 @@ impl<'env> Mempool<'env> {
...
+                self.state.write().filter.blacklist( &transaction);
```
Comment (paberr) [inline mempool/src/lib.rs]: Unnecessary space.
```suggestion
                self.state.write().filter.blacklist(&transaction);
```

File: mempool/src/lib.rs [hunk 12]
```diff
@@ -126,8 +139,16 @@ impl<'env> Mempool<'env> {
             if recipient_account.account_type() != transaction.recipient_type {
                 return ReturnCode::Invalid;
             }
-            if let Err(_) = recipient_account.with_incoming_transaction(&transaction, block_height) {
-                return ReturnCode::Invalid;
+
+            match recipient_account.with_incoming_transaction(&transaction, block_height) {
+                Err(_) => return ReturnCode::Invalid,
+                Ok(r) => {
+                    // Check recipient account against filter rules
```
Comment (paberr) [inline mempool/src/lib.rs]: Most of our comments end with a `.`.
```suggestion
                    // Check recipient account against filter rules.
```

File: mempool/src/lib.rs [hunk 13]
```diff
@@ -556,7 +582,8 @@ pub enum ReturnCode {
     FeeTooLow,
     Invalid,
     Accepted,
-    Known
+    Known,
+    Filtered
```
Comment (paberr) [inline mempool/src/lib.rs]: For easy extendibility, add a `,`.
```suggestion
    Filtered,
```

File: mempool/src/lib.rs [hunk 14]
```diff
@@ -157,9 +178,14 @@ impl<'env> Mempool<'env> {
                     break;
                 }
 
+                let old_sender_account = sender_account.clone();
                 sender_account = sender_account
                     .with_outgoing_transaction(tx, block_height)
                     .expect("Failed to apply existing transaction");
+                if !state.filter.accepts_sender_account(&transaction, &old_sender_account, &sender_account) {
```
Comment (paberr) [inline mempool/src/lib.rs]: I think this is the wrong place to do your check.
In this loop, we are applying other transactions.
You might want to move your check towards line ~200 (see my comment there).

Comment (paberr) [reply mempool/src/lib.rs]: Ok, I cannot add comments in lines that haven't changed.
It is around `// Now, check the new transaction.` when we actually apply the current transaction to the Mempool.

**Review by paberr — CHANGES_REQUESTED** [review]

**Review by paberr — APPROVED** [review]

File: mempool/src/filter.rs (hunk 5 above)
Comment (jgraef) [reply mempool/src/filter.rs]: For short constructors, I usually also format them in one line. Do we have a concrete code style? I think Rust suggest a certain code style and we could use auto-format to enforce it. Anyway, maybe we should discuss this in a seperate Issue.

**Pull request merged**